### ISSUE TITLE
API 'QString' has already been set to version 1

### ISSUE THREAD

**Issue by JeroenDelcour** [PierreRaybaut/guidata#35]
Trying to run [roibuddy](https://github.com/losonczylab/roibuddy) results in the following error message:

```
/home/jeroen.delcour/anaconda/lib/python2.7/site-packages/skimage/filter/__init__.py:6: skimage_deprecation: The `skimage.filter` module has been renamed to `skimage.filters`.  This placeholder module will be removed in v0.13.
  warn(skimage_deprecation('The `skimage.filter` module has been renamed '
Traceback (most recent call last):
  File "/home/jeroen.delcour/anaconda/bin/roibuddy", line 7, in <module>
    from roibuddy.roi_buddy import main
  File "/home/jeroen.delcour/anaconda/lib/python2.7/site-packages/roibuddy/roi_buddy.py", line 32, in <module>
    from guidata import qthelpers
  File "/home/jeroen.delcour/anaconda/lib/python2.7/site-packages/guidata/__init__.py", line 541, in <module>
    import guidata.config
  File "/home/jeroen.delcour/anaconda/lib/python2.7/site-packages/guidata/config.py", line 15, in <module>
    from guidata.configtools import add_image_module_path, get_translation
  File "/home/jeroen.delcour/anaconda/lib/python2.7/site-packages/guidata/configtools.py", line 22, in <module>
    from guidata.qt.QtGui import (QFont, QLabel, QPixmap, QIcon, QHBoxLayout,
  File "/home/jeroen.delcour/anaconda/lib/python2.7/site-packages/guidata/qt/__init__.py", line 25, in <module>
    sip.setapi('QString', 2)
ValueError: API 'QString' has already been set to version 1
```

Running all latest versions of roibuddy (1.0.0), guidata (1.7.4), pythonqwt (0.5.4), and spyder (2.3.8). Using Anaconda 2.3.0, Python 2.7.11, 64-bit.
Have also tried using a fresh install of Anaconda on Windows, resulting in exactly the same error.

Perhaps Spyder is the one to have already set `API 'QString'` to 1?

**Comment by PierreRaybaut** [comment]
The latest version of guidata is based on PyQt API 2. My guess is that Spyder is probably setting PyQt API to 1 before guidata initializes itself.

Could you check that it works outside Spyder? (i.e. from a terminal)
Inside Spyder, options related to PyQt are available in "Console" section, if I remember correctly: you may choose to switch to PyQt API 2 there.

**Comment by JeroenDelcour** [comment]
That error was actually produced by running from a terminal. I'll check out the PyQt settings in Spyder and see if I can get it to run by changing that, thanks.

**Comment by nobias** [comment]
Hi,
same problem here, I was able to narrow it down a little. Doesn't seem to be related to Spyder in any way. In an up-to-date Anaconda on Scientific Linux 6.6, in a plain, fresh ipython (i.e., no qtconsole or anything), if I import guidata.qthelpers after PyQt4 (like it is done in roibuddy):
```
In [1]: from PyQt4.QtCore import *

In [2]: from PyQt4.QtGui import *

In [3]: from guidata import qthelpers
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-0cce8cd243ee> in <module>()
----> 1 from guidata import qthelpers

/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/guidata/__init__.py in <module>()
    539 #    Version 2.0 dated 2006-09-05.
    540 
--> 541 import guidata.config
    542 
    543 def qapplication():

/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/guidata/config.py in <module>()
     13 import os.path as osp
     14 
---> 15 from guidata.configtools import add_image_module_path, get_translation
     16 from guidata.userconfig import UserConfig
     17 

/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/guidata/configtools.py in <module>()
     20 import gettext
     21 
---> 22 from guidata.qt.QtGui import (QFont, QLabel, QPixmap, QIcon, QHBoxLayout,
     23                               QColor, QPen, QBrush, QFontDatabase)
     24 from guidata.qt.QtCore import Qt

/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/guidata/qt/__init__.py in <module>()
     23     import sip
     24     try:
---> 25         sip.setapi('QString', 2)
     26         sip.setapi('QVariant', 2)
     27         sip.setapi('QDate', 2)

ValueError: API 'QString' has already been set to version 1
```

If I import guidata.qthelpers before PyQt, then QString and other clases in PyQt are not defined:

```
In [1]: from guidata import qthelpers

In [2]: from PyQt4.QtCore import *

In [3]: from PyQt4.QtGui import *

In [4]: QString
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-4-900c7a89f8ae> in <module>()
----> 1 QString

NameError: name 'QString' is not defined
```
Any clues? My PyQt is at 4.11.4, guidata at 1.7.4.

**Comment by stonebig** [comment]
hi,

no issue on WinPython-64bit-3.4.4.1build1, trying with ptpython:
![guidata2016-02](https://cloud.githubusercontent.com/assets/4312421/12307688/1ea6d3fc-ba41-11e5-9f2e-4db4d31a0648.GIF)

**Comment by stonebig** [comment]
no issue with Spyder 3.0 dev ot qt console either:
![guidata2016-03](https://cloud.githubusercontent.com/assets/4312421/12307805/e8172a0c-ba41-11e5-8f67-0df357e66a21.GIF)
![guidata2016-01](https://cloud.githubusercontent.com/assets/4312421/12307810/ecf8cde6-ba41-11e5-8c9e-9269c0bc99a7.GIF)

**Comment by PierreRaybaut** [comment]
The fact that QString is no longer defined in PyQt4.QtCore after having imported guidata v1.7 is normal: guidata is using PyQt API 2 in which QString has been replaced by Python strings (PyQt API 2 is the current official API on Python 3, it’s the future).




But there should not be the “API 'QString' has already been set to version 1 ” error… have you checked your environment variables? (IPython must import PyQt before guidata and switch to API 1, based on an env var for example)


More importantly, please do check this with a real Python interpreter (IPython is a tweaked Python interpreter: it does a lot of things, sometimes unwanted, before running your commands): simply type “python” in a terminal.





De : Tobias Nöbauer
Envoyé : ‎mercredi‎ ‎13‎ ‎janvier‎ ‎2016 ‎20‎:‎59
À : PierreRaybaut/guidata
Cc : Pierre Raybaut





Hi,
same problem here, I was able to narrow it down a little. Doesn't seem to be related to Spyder in any way. In an up-to-date Anaconda on Scientific Linux 6.6, in a plain, fresh ipython (i.e., no qtconsole or anything), if I import guidata.qthelpers after PyQt4 (like it is done in roibuddy):
In [1]: from PyQt4.QtCore import *

In [2]: from PyQt4.QtGui import *

In [3]: from guidata import qthelpers
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-0cce8cd243ee> in <module>()
----> 1 from guidata import qthelpers

/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/guidata/__init__.py in <module>()
    539 #    Version 2.0 dated 2006-09-05.
    540 
--> 541 import guidata.config
    542 
    543 def qapplication():

/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/guidata/config.py in <module>()
     13 import os.path as osp
     14 
---> 15 from guidata.configtools import add_image_module_path, get_translation
     16 from guidata.userconfig import UserConfig
     17 

/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/guidata/configtools.py in <module>()
     20 import gettext
     21 
---> 22 from guidata.qt.QtGui import (QFont, QLabel, QPixmap, QIcon, QHBoxLayout,
     23                               QColor, QPen, QBrush, QFontDatabase)
     24 from guidata.qt.QtCore import Qt

/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/guidata/qt/__init__.py in <module>()
     23     import sip
     24     try:
---> 25         sip.setapi('QString', 2)
     26         sip.setapi('QVariant', 2)
     27         sip.setapi('QDate', 2)

ValueError: API 'QString' has already been set to version 1


If I import guidata.qthelpers before PyQt, then QString and other clases in PyQt are not defined:
In [1]: from guidata import qthelpers

In [2]: from PyQt4.QtCore import *

In [3]: from PyQt4.QtGui import *

In [4]: QString
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-4-900c7a89f8ae> in <module>()
----> 1 QString

NameError: name 'QString' is not defined


Any clues? My PyQt is at 4.11.4, guidata at 1.7.4.

**Comment by PierreRaybaut** [comment]
And by the way, like @stonebig, I'm unable to reproduce this issue with WinPython even with Python 2.7.10 in which PyQt default API is API 2 (inside or outside Spyder).

**Comment by stonebig** [comment]
would it be possible that "base" Anaconda launcher is doing something bad ?

**Comment by nobias** [comment]
Exact same problem in bare Python interpreter (from Anaconda). Below my environment on that linux box:
```
>>> print os.environ
{'_': '/home/tobias.noebauer/opt/anaconda/bin/python', 'XAUTHORITY': '/home/tobias.noebauer/.Xauthority', 'LOADEDMODULES': '', 'KDE_IS_PRELINKED': '1', 'GNOME_DESKTOP_SESSION_ID': 'this-is-deprecated', 'QT_GRAPHICSSYSTEM': 'native', 'MATLABPATH': '/home/tobias.noebauer/.matlab/my_startup:/home/tobias.noebauer/devel/bfmatlab', 'X2GO_AGENT_PID': '6921', 'LESSOPEN': '||/usr/bin/lesspipe.sh %s', 'SSH_CLIENT': '10.42.34.251 62249 22', 'QT_IM_MODULE': 'xim', 'LOGNAME': 'tobias.noebauer', 'USER': 'tobias.noebauer', 'PATH': '/opt/matlab/bin:/home/tobias.noebauer/opt/anaconda/bin:/usr/local/cuda-7.5/bin:/home/tobias.noebauer/bin:/usr/libexec/nx:/opt/matlab/bin:/home/tobias.noebauer/opt/anaconda/bin:/usr/local/cuda-7.5/bin:/home/tobias.noebauer/bin:/usr/libexec/nx:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/lib64/mpich/lib:/usr/lib64/mpich/lib', 'QTDIR': '/usr/lib64/qt-3.3', 'GTK_IM_MODULE': 'gtk-im-context-simple', 'LD_LIBRARY_PATH': '/opt/matlab/bin/glnxa64:/opt/matlab/sys/os/glnxa64:/usr/lib64/nx/X11/Xinerama:/usr/lib64/nx/X11:/opt/matlab/bin/glnxa64:/opt/matlab/sys/os/glnxa64::/usr/local/cuda-7.5/lib64:/usr/local/cuda-7.5/lib64', 'SSH_AGENT_PID': '7298', 'LANG': 'en_US.UTF-8', 'HISTCONTROL': 'ignoredups', 'TERM': 'xterm', 'SHELL': '/bin/bash', 'XDG_SESSION_COOKIE': '35aa7dbe1d5482cb645431a500000025-1452081110.360642-1393581270', 'SESSION_MANAGER': 'local/unix:@/tmp/.ICE-unix/7297,unix/unix:/tmp/.ICE-unix/7297', 'QTINC': '/usr/lib64/qt-3.3/include', 'G_BROKEN_FILENAMES': '1', 'SITK_SHOW_COMMAND': 'imagej -eval \'open( "%f" );\'', 'HISTSIZE': '1000', 'WINDOWID': '39845891', 'XSESSION_EXEC': '/usr/bin/gnome-session', 'IMSETTINGS_INTEGRATE_DESKTOP': 'yes', 'HOME': '/home/tobias.noebauer', 'MODULESHOME': '/usr/share/Modules', 'NX_XINERAMA_CONF': '/home/tobias.noebauer/.x2go/C-tobias.noebauer-50-1452081102_stDGNOME_dp32/xinerama.conf', 'IMSETTINGS_MODULE': 'none', 'MSM_PRODUCT': 'MSM', 'COLORTERM': 'gnome-terminal', 'QTLIB': '/usr/lib64/qt-3.3/lib', 'PYTHONPATH': '/home/tobias.noebauer/opt/anaconda/lib/python2.7/site-packages/', 'SSH_AUTH_SOCK': '/tmp/keyring-L0TvJS/socket.ssh', 'XMODIFIERS': '@im=none', 'BASH_FUNC_module()': '() {  eval `/usr/bin/modulecmd bash $*`\n}', 'DBUS_SESSION_BUS_ADDRESS': 'unix:abstract=/tmp/dbus-AFMLo5cNue,guid=597f73a710c1c91e621d3caa0000014c', 'ORBIT_SOCKETDIR': '/tmp/orbit-tobias.noebauer', 'MODULEPATH': '/usr/share/Modules/modulefiles:/etc/modulefiles', 'KDEDIRS': '/usr', 'GNOME_KEYRING_SOCKET': '/tmp/keyring-L0TvJS/socket', 'SSH_CONNECTION': '10.42.34.251 62249 10.42.158.34 22', 'DISPLAY': ':50.0', 'MSM_HOME': '/usr/local/MegaRAID Storage Manager', 'OLDPWD': '/home/tobias.noebauer/build/sima-0.2.1', 'HOSTNAME': 'pcl-imp-2.imp.univie.ac.at', 'STARTUP': '/usr/bin/env LD_LIBRARY_PATH=/usr/lib64/nx/X11/Xinerama:/usr/lib64/nx/X11:/opt/matlab/bin/glnxa64:/opt/matlab/sys/os/glnxa64::/usr/local/cuda-7.5/lib64 /usr/bin/gnome-session', 'SITK_SHOW_COLOR_COMMAND': 'imagej -eval \'open( "%f" );\'', 'SHLVL': '5', 'PWD': '/home/tobias.noebauer', 'CVS_RSH': 'ssh', 'GTK_RC_FILES': '/etc/gtk/gtkrc:/home/tobias.noebauer/.gtkrc-1.2-gnome2', 'MAIL': '/var/spool/mail/tobias.noebauer', 'LS_COLORS': 'rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=01;05;37;41:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lz=01;31:*.xz=01;31:*.bz2=01;31:*.tbz=01;31:*.tbz2=01;31:*.bz=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.rar=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=01;36:*.au=01;36:*.flac=01;36:*.mid=01;36:*.midi=01;36:*.mka=01;36:*.mp3=01;36:*.mpc=01;36:*.ogg=01;36:*.ra=01;36:*.wav=01;36:*.axa=01;36:*.oga=01;36:*.spx=01;36:*.xspf=01;36:', 'X2GO_SESSION': 'tobias.noebauer-50-1452081102_stDGNOME_dp32'}
```
Nothing too strange, I would say. Also, keep in mind that we are seeing the same issue when using a fresh Anaconda on Windows. So Anaconda seems to cause this, in whatever way.

**Comment by PierreRaybaut** [comment]
Could you please copy/paste your environment variables here?

**Comment by PierreRaybaut** [comment]
I was expected something related with Qt API in your environment variables but that's not the case.

Oh! wait a minute, @nobias: I think I've read your first message too fast!

Remember that the default PyQt API depends on Python version:
  - Python 2.7: PyQt API 1 (old API with QString, ...)
  - Python 3 : PyQt API 2 (new pythonic API with unicode replacing QString, ...)

So, with Python 2.7, if you import PyQt, the PyQt API 1 will be selected by default, unless you do the following first:
```python
import sip
sip.setapi('QString', 2)
[...]
```

Let's illustrate how it works with the following code snippets:
  - Python 2.7:
    - Default PyQt API is API 1
    - You must either import `guidata` before importing PyQt or any other library based on PyQt
```python
## Importing PyQt before importing any other PyQt-based library 
## will do implicitely: `import sip ; sip.setapi('QString', 1)`
from PyQt4.QtCore import *
from PyQt4.QtGui import *
# At this point, PyQt API 1 has been selected
from guidata import qthelpers
# This will raise the exception:
# "ValueError: API 'QString' has already been set to version 1"
# Why? because you can't change PyQt API after it has been selected 
# and PyQt has been imported.
# So, that is the expected behavior.
# To prevent this error, just import guidata first, that's all.
```
  - Python 3.4: importing PyQt will select PyQt API 2
    - Default PyQt API is API 2
    - Importing `guidata` will work even if you import first PyQt or any other library based on PyQt
```python
## Importing PyQt before importing any other PyQt-based library 
## will do implicitely: `import sip ; sip.setapi('QString', 2)`
from PyQt4.QtCore import *
from PyQt4.QtGui import *
# At this point, PyQt API 2 has been selected
from guidata import qthelpers
# This will raise no exception because it's based on PyQt API 2.
```

So, in other words, **the behavior reported by @nobias is normal.**

**Comment by PierreRaybaut** [comment]
@JeroenDelcour: in your case, I think that [roibuddy](https://github.com/losonczylab/roibuddy) is not compatible with latest version of `guidata` (1.7) and `guiqwt` (3.0). To make it compatible, you will have to import `guidata.qt` first:
```python
from guidata import qt   # import guidata.qt will select PyQt API 2 (required on Python 2.7 only)
import roibuddy
```

I see that an issue has already been opened on [roibuddy project page](https://github.com/losonczylab/roibuddy): https://github.com/losonczylab/roibuddy/issues/21.

Please confirm that it works so I can close this issue.

**Comment by PierreRaybaut** [comment]
I've proposed some changes in roibuddy to help fixing this compatiblity issue: https://github.com/losonczylab/roibuddy/pull/23.

**Comment by nobias** [comment]
OK, I see. Sorry for my ignorance about Python and Qt API versions. I just tested your pull request to roibuddy and reported on the result there: losonczylab/roibuddy#23: Works for me with one minor change.
Thanks for your great support and quick reaction!
Tobias

**Comment by JeroenDelcour** [comment]
Works for me too, with @nobias his minor. Thanks for the extremely fast and helpful response! :)

**Comment by jmamuedo** [comment]
PierreRaybaut.

I see you have a deep knowledge about PyQt API and Python. So I have a question to resolve my problem. I'm trying use 'pyuic4' to compile ui file to py file. But I recive this message:

 `usuario@usuario:~/.qgis2/python/plugins/MyPlugin$ pyuic4 myplugin_dialog_base.ui -o myplugin_dialog.py -d

Traceback (most recent call last):
 File "/usr/lib/python2.7/dist-packages/PyQt4/uic/port_v2/invoke.py", line 39, in invoke
   exit_status = driver.invoke()
 File "/usr/lib/python2.7/dist-packages/PyQt4/uic/driver.py", line 64, in invoke
   self._generate()
 File "/usr/lib/python2.7/dist-packages/PyQt4/uic/driver.py", line 99, in _generate
   self._opts.resource_suffix)
 File "/usr/lib/python2.7/dist-packages/PyQt4/uic/__init__.py", line 175, in compileUi
   winfo = compiler.UICompiler().compileUi(uifile, pyfile, from_imports, resource_suffix)
 File "/usr/lib/python2.7/dist-packages/PyQt4/uic/Compiler/compiler.py", line 55, in __init__
   CompilerCreatorPolicy())
 File "/usr/lib/python2.7/dist-packages/PyQt4/uic/uiparser.py", line 145, in __init__
   self.factory = QObjectCreator(creatorPolicy)
 File "/usr/lib/python2.7/dist-packages/PyQt4/uic/objcreator.py", line 94, in __init__
   modinfo = plugin_locals["moduleInformation"]()
 File "<string>", line 52, in moduleInformation
 File "/usr/lib/python2.7/dist-packages/qgis/__init__.py", line 26, in <module>
   from qgis.PyQt import QtCore
 File "/usr/lib/python2.7/dist-packages/qgis/PyQt/QtCore.py", line 28, in <module>
   sip.setapi(api, 2)
ValueError: API 'QDate' has already been set to version 1`

So my case is Python 2.7: Default PyQt API is API 1. What items do I have to change to a correctly working? I'm new in this topic and I have not experience. 

Thanks.

**Comment by Zac-Mwangi** [comment]
i have this error while opening tuxcut.. 
any suggestions???

<PyQt4.QtCore.QVariant object at 0x7febd57a32a8>
Traceback (most recent call last):
  File "/opt/TuxCut/run.py", line 18, in <module>
    tux = TuxCut()
  File "/opt/TuxCut/TuxCut.py", line 15, in __init__
    uic.loadUi('ui/MainWindow.ui',self)
  File "/usr/lib/python2.7/dist-packages/PyQt4/uic/__init__.py", line 236, in loadUi
    return DynamicUILoader(package).loadUi(uifile, baseinstance, resource_suffix)
  File "/usr/lib/python2.7/dist-packages/PyQt4/uic/Loader/loader.py", line 50, in __init__
    UIParser.__init__(self, QtCore, QtGui, LoaderCreatorPolicy(package))
  File "/usr/lib/python2.7/dist-packages/PyQt4/uic/uiparser.py", line 183, in __init__
    self.factory = QObjectCreator(creatorPolicy)
  File "/usr/lib/python2.7/dist-packages/PyQt4/uic/objcreator.py", line 94, in __init__
    modinfo = plugin_locals["moduleInformation"]()
  File "<string>", line 52, in moduleInformation
  File "/usr/lib/python2.7/dist-packages/qgis/__init__.py", line 26, in <module>
    from qgis.PyQt import QtCore
  File "/usr/lib/python2.7/dist-packages/qgis/PyQt/QtCore.py", line 28, in <module>
    sip.setapi(api, 2)
ValueError: API 'QDate' has already been set to version 1
